File: data-structures/strings/javascript/stringOperations.js
/**
 * declaration / initialization
 */
let str1 = "This is a string";
let str2 = 'This is also a string';
let str3 = `Template literal string, ${str1}`;

/**
 * accessing characters
 */
let char = str1[5];
console.log(char); // 5 

/**
 * string length
 */
let len = str1.length;
console.log(len);   // 16

/**
 * finding a substring
 */
// first occurrence, -1 if not found - 17
let position = str3.indexOf("string");
console.log(position); 
// last occurrence - 35
let lastPosition = str3.lastIndexOf("string");
console.log(lastPosition);
// checking substring existence - true
let exists = str1.includes("string");
console.log(exists);

/**
 * substring, substr, slice
 */
// substring(start, end);
// extract characters from start to (end - 1) - 'is a s'
let sub = str1.substring(5, 11);
console.log(sub);
// slice(start, end);
// similar to substring but can accept negative indices - 'is a strin'
let sub2 = str1.slice(5, -1);
console.log(sub2);

/**
 * string manipulation
 */
// concatenation
let concatedStr = str1 + " " + str2;
console.log(concatedStr);
let concatedStr2 = str1.concat(str2);
console.log(concatedStr2);
// template literals
let name = "Matt";
let greeting = `Hello ${name}!`;
console.log(greeting);
// changing case
let upper = str1.toUpperCase();
console.log(upper);
let lower = str1.toLowerCase();
console.log(lower);
// trimming spaces at start and end
let trimmed = "    extra spaces     ".trim();
console.log(trimmed);

/**
 * splitting and joining
 */
// split(delimiter); - string into an array
let words = str1.split(" ");
console.log(words);
// join(delimiter); - array into string
let sentence = words.join(" ");
console.log(sentence);

/**
 * replacing text
 */
let replaced = str1.replace("string", "replacedString");
console.log(replaced);

File: data-structures/strings/java/StringOperations.java
// Java strings are immutable hence any modification leads to a new string. 
// Java uses a String pool where literal strings are stored for memory optimization.

public class StringOperations {
    public static void main(String[] args){
        /**
         * declaration / initialization
         */
        // this uses the String pool
        String str1 = "hello";

        // this does not use the String pool
        String str2 = new String("hello");

        // concatenation
        String str3 = str1 + " " + str2;
        System.out.println(str3);               // "hello hello"

        /**
         * methods and operations
         */
        // length of string
        int len = str3.length();
        System.out.println(len);                // 11

        // charAt
        char firstLetter = str3.charAt(0);
        System.out.println(firstLetter);        // h

        // substring
        String sub = str3.substring(1, 3);
        System.out.println(sub);                // el

        // split - ["Hello", 'World']
        String[] partsArr = "Hello, World".split(", ");
        for (String part: partsArr){
            System.out.println(part);
        }
        
        // replace
        String replaced = str1.replace("e", "a");
        System.out.println(replaced);           // hallo

        // case conversion
        String upper = str1.toUpperCase();
        String lower = str1.toLowerCase();
        System.out.println(upper);              // HELLO
        System.out.println(lower);              // hello

        // trim whitespace
        String trimmed = "    hello     ".trim();
        System.out.println(trimmed);            // "hello"

        // starts / ends with
        boolean starts = str1.startsWith("he");
        System.out.println(starts);             // true
        boolean ends = str2.endsWith("lo");
        System.out.println(ends);               // true

        // indexOf / lastIndexof
        int pos = str1.indexOf('l');
        System.out.println(pos);                // 2
        int lastPos = str1.lastIndexOf('l');
        System.out.println(lastPos);            // 3

        /**
         * comparison
         */
        // compare content 
        boolean isEqual = str1.equals(str2);
        System.out.println(isEqual);            // true

        // compare case-insensitive content 
        boolean isEqualIgnoreCase = str1.equalsIgnoreCase("HELLO");
        System.out.println(isEqualIgnoreCase);  // true

        // lexicographical comparison - 0, 7, -17
        int comparison = str1.compareTo(str2);
        System.out.println(comparison);         // 0
        comparison = str1.compareTo("allo");
        System.out.println(comparison);         // 7
        comparison = str1.compareTo("yello");
        System.out.println(comparison);         // 17

        /**
         * StringBuilder(non-syncronized) & StringBuffer(syncronized)
         */
        // are more efficient and mutable 
        StringBuilder sb = new StringBuilder();
        sb.append("Hello");
        System.out.println(sb);         // Hello
        sb.append(" ");
        sb.append("world");
        String result = sb.toString();
        System.out.println(result);     // Hello world

        /**
         * conversion
         */
        // string to char array 
        char[] chars = str1.toCharArray();
        System.out.print("[ ");
        for (char c:chars){
            System.out.print(c + ", ");
        }
        System.out.println(']');                // [ h, e, l, l, o, ]

        // string to integer 
        int number = Integer.parseInt("123");
        System.out.println(number);               // 123

        // integer to string 
        String numString = Integer.toString(123);
        System.out.println(numString);            // "123"
    }
}